services/web-server/test/pulseengine_test.js

```diff
@@ -0,0 +1,153 @@
+  // does not activate immediately
+  const beat = () => new Promise(resolve => setTimeout(resolve, 1));
+
+  it('can be constructed with a fake client', function() {
+    // can't test much more of this since PulseEngine accesses amqplib APIs directly
+    new PulseEngine({monitor: manager.monitor(), pulseClient: new FakeClient()});
+  });
+
+  describe('PulseIterator', function() {
+    it('queues up pushed messages', async function() {
+      const engine = new FakePulseEngine();
+      const pi = new PulseIterator(engine, []);
+
+      assert.equal(engine.subscribed, true, "should have subscribed");
+
+      // first push a bunch of messages..
+      const sent = [];
+      ['M1', 'M2', 'M3'].forEach((msg) => {
+        engine.onMessage(msg).then(() => sent.push(msg), err => sent.push(err));
+      });
+
+      // messages aren't sent yet..
+      await beat();
+      assert.deepEqual(sent, []);
+
+      // now see that those appear in the output
+      const received = [];
+      for await (let msg of pi) {
+        received.push(msg);
+        if (received.length === 3) {
+          break;
+        }
+      }
+
+      assert.equal(engine.subscribed, false, "should have been unsubscribed");
+      assert.deepEqual(sent, ['M1', 'M2', 'M3']);
+      assert.deepEqual(received, ['M1', 'M2', 'M3']);
+    });
+
+    it('waits for pushed messages', async function() {
+      const engine = new FakePulseEngine();
+      const pi = new PulseIterator(engine, []);
+
+      // start waiting for output before messages arrive..
+      const result = [];
+      const finished = (async () => {
+        for await (let msg of pi) {
+          result.push(msg);
+          if (result.length === 3) {
+            break;
+          }
+        }
+      })();
+
+      await beat();
+      assert.deepEqual(result, []);
+
+      await engine.onMessage('M1');
+      await beat();
+      assert.deepEqual(result, ['M1']);
+
+      await engine.onMessage('M2');
+      await beat();
+      assert.deepEqual(result, ['M1', 'M2']);
+
+      await engine.onMessage('M3');
+      await beat();
+      assert.deepEqual(result, ['M1', 'M2', 'M3']);
+
+      await finished;
+
+      assert.equal(engine.subscribed, false, "should have been unsubscribed");
+    });
+
+    it('throws errors into the iterator', async function() {
+      const engine = new FakePulseEngine();
+      const pi = new PulseIterator(engine, []);
+
+      const sent = [];
+      engine.onMessage('M1').then(() => sent.push('M1'));
+      assert.deepEqual(sent, []);
+      assert.deepEqual(await pi.next(), {value: 'M1', done: false});
+      assert.deepEqual(sent, ['M1']);
+
+      // an unsent M2 will error out after onError is called
+      engine.onMessage('M2')
+        .then(() => { throw new Error('unexpected success'); })
+        .catch(err => sent.push(err.toString()));
+      assert.deepEqual(sent, ['M1']);
+
+      engine.onError(new Error('uhoh'));
```

Comment:
Should this listener be set up before L117?

Reply:
It's not a listener -- it's a callback.  The name is misleading.  I'll change `onError` and `onMessage` to be `handleXxx` isntead.